### PR TITLE
Fix torch.gcd returning negative results for INT_MIN inputs

### LINKED ISSUE #187338: torch.gcd returns an incorrect negative result when one input contains the minimum value of torch.int8 (-128)
### Description
`torch.gcd` returns an incorrect negative result when one input contains the minimum value of `torch.int8` (-128).
For example, gcd(-128, 124) should be 4, but PyTorch returns -4 when both tensors have dtype torch.int8.
This seems to be related to handling the minimum signed integer value, whose absolute value cannot be represented in the same signed integer dtype.

### Reproduction
```python
import torch
import numpy as np

x = torch.tensor([124, -128, -126], dtype=torch.int8)
y = torch.tensor([124, 124, 126], dtype=torch.int8)
print(torch.gcd(x, y))

x_np = np.array([124, -128, -126], dtype=np.int8)
y_np = np.array([124, 124, 126], dtype=np.int8)
print(np.gcd(x_np, y_np))
```

### Output
```
tensor([124,  -4, 126], dtype=torch.int8)
[124   4 126]
```

### Versions

Versions of relevant libraries:
[pip3] numpy==2.2.6
[pip3] optree==0.19.1
[pip3] torch==2.12.0

### PR DESCRIPTION
**Pull request by razaaliraza**
`torch.gcd` returns incorrect negative results when one input is the minimum value of a signed integer type. For example, `gcd(-128, 124)` returns `-4` instead of `4` for int8.

`calc_gcd` calls `std::abs()` on inputs, which is undefined behavior for INT_MIN because the positive result overflows. The fix uses unsigned arithmetic instead: cast to `std::make_unsigned_t<T>` before negation, which is well-defined.

Fixed on CPU and CUDA (precompiled and Jiterator). The Jiterator path requires an inline `make_unsigned_impl` trait since the JIT compilation environment does not include `<type_traits>`.

Edge case: `gcd(INT_MIN, 0)` returns INT_MIN (negative) because the true result is not representable in the signed type. This matches NumPy.

Fixes #187338

cc @albanD

### CODE REVIEW

**Comment by malfet** [comment]
> @pytorchbot label "topic: not user facing"

If it's not user facing, then why are you adding a new test for it?

**Comment by malfet** [comment]
@claude review this

**Comment by razaaliraza** [comment]
> > @pytorchbot label "topic: not user facing"
> 
> If it's not user facing, then why are you adding a new test for it?

Sure, I'll take that note.

**Comment by razaaliraza** [comment]
Also if you want cuda changes to be part of this PR, let me know and I will ad those. Cuda and MPS both need changes for this, but I am sure MPS will have to be a different PR. Also LCM calculation suffers from this but I didnt want to cram too much in a single PR. Definitely we can do cpu and cuda together. Let me know, thanks!

**Comment by razaaliraza** [comment]
@malfet fixed all the comments raised by claude. Updated PR desc. Have a look. Thanks!

**Comment by razaaliraza** [comment]
rebased on viable/strict and fixed conflicts. Please have a look. Thanks! @malfet
